pimd/pim_pim.c

```diff
@@ -138,31 +138,33 @@ void pim_sock_delete(struct interface *ifp, const char *delete_message)
 }
 
 /* For now check dst address for hello, assrt and join/prune is all pim rtr */
-static bool pim_pkt_dst_addr_ok(enum pim_msg_type type, in_addr_t addr)
+static bool pim_pkt_dst_addr_ok(enum pim_msg_type type, pim_addr addr)
 {
 	if ((type == PIM_MSG_TYPE_HELLO) || (type == PIM_MSG_TYPE_ASSERT)
 	    || (type == PIM_MSG_TYPE_JOIN_PRUNE)) {
-		if (addr != qpim_all_pim_routers_addr.s_addr)
+		if (pim_addr_cmp(addr, qpim_all_pim_routers_addr))
 			return false;
 	}
 
 	return true;
 }
 
-int pim_pim_packet(struct interface *ifp, uint8_t *buf, size_t len)
+int pim_pim_packet(struct interface *ifp, uint8_t *buf, size_t len,
+		   pim_sgaddr sg)
 {
-	struct ip *ip_hdr;
+#if PIM_IPV == 4
+	struct ip *ip_hdr = (struct ip *)buf;
 	size_t ip_hlen; /* ip header length in bytes */
-	char src_str[INET_ADDRSTRLEN];
-	char dst_str[INET_ADDRSTRLEN];
+#endif
 	uint8_t *pim_msg;
-	int pim_msg_len;
+	uint32_t pim_msg_len = 0;
 	uint16_t pim_checksum; /* received checksum */
```

Comment:
For IPv6 the checksum in PIM header has to be computed including a pseudo header. Refer section 4.9 in RFC7761. Checksum validation should include IPv6 Pseudoheader.

Reply:
Indeed, I'll add it to this PR.